### ISSUE TITLE
SNES cores display wrong PAR with "crop overscan" OFF.

### ISSUE THREAD

**Issue by vaguerant** [libretro/RetroArch#2126]
Some background on SNES resolution and overscan comes first, skip if you know this stuff.

**Background:** On the NES, all software ran at 256*240. NTSC, PAL, everything is 240 lines. Occasionally, developers would run into issues by dropping things on the first or last few lines which they expected the player to be able to see. Thus, they would have to explicitly display nothing (important) in the overscan area (at least lines 0-7 and 232-239 are a sensible amount to cut, leaving the 224 lines in the vertical center for your game) and only start drawing at line 8 or so, stopping again at line 231 if necessary.

On the Super Famicom/SNES, the (standard) maximal resolution is 256*239, but Nintendo implemented a feature where developers could simply run the console in a 224-line mode. They would write lines 0-223 to the framebuffer, set the 224-line mode on the SNES, and the SNES would throw in the (vertical) overscan automatically. Lines 224-238 can simply be ignored by the developer. Of course, the video output by the SNES is still 240p (let's ignore PAL), but in most practical senses, the software itself is running at 224 lines while the SNES handles overscan.

Of course, a handful of titles did use all 239 lines. This works exactly as you'd expect; on a standard TV, you'll likely lose a significant portion of the picture around the top and bottom edges. So there are only a few NTSC titles running at this resolution, and then Nintendo were nice enough to expand some of their first-party SNES games to use 239 lines on PAL machines, to reduce the massive letterboxing by a small degree (239/288 looks better than 224/288).

Anyway, getting back to 224-line games: if you were to force display of all 239 lines the SNES hardware is capable of, as some emulators are able to, your picture would be at the top of the video frame with a (generally) empty bar at the bottom (the unused last 15 lines). However, most or possibly all emulators will only bother to display the resolution that the game uses--i.e., they're not emulating the SNES video signal, they're just showing what's important: the lines which the game was written with the expectation players would see--so 99% of the time SNES emulators are in use, they'll be outputting a picture with 224 lines and no overscan.

**The RetroArch issue:** I've tested several of the SNES cores (Snes9x, Snes9x-Next, bsnes), and when "crop overscan" is set to ON, the results are as following:
* 224-line games have a 224-line picture
* 239-line games are cropped to have a 224-line picture
* setting 1:1 PAR in the video options results in an 8:7, 1:1 PAR, 256*224 picture in both cases

That's all correct and expected behavior. The issue appears when "crop overscan" is set to OFF, where the results are as following:
* 224-line games have a 224-line picture
* 239-line games have a 239-line picture
* **setting 1:1 PAR in the video options results in a 256:239 picture in both cases**

RetroArch appears to expect that when "crop overscan" is OFF, SNES cores will output a 256\*239 picture, but this is **false**. 224-line games continue to run at 224 lines, as there are no additional lines to "uncrop" being fed to RetroArch--remember, emulators are only displaying the lines from the mode the software is running, so 224 lines in 224-line titles, 239 lines in 239-line titles.

Basically, the PAR **cannot** be correctly judged by checking whether "crop overscan" is turned off. When "crop overscan" is ON, RetroArch **can** reasonably assume that substantially all (let's ignore the couple of 448-line titles right now) SNES games will be feeding it 224 lines to display--either all lines or the middle 224 lines of a 239-line title.

However, when "crop overscan" is OFF, RetroArch **cannot** be sure whether the image fed to it will be 224 lines or 239 lines. The resolution must not be determined using this method alone because it will be wrong in almost all cases. To display correctly, the method of determining the number of lines displayed must be based on something other than "crop overscan".

**Comment by fr500** [comment]
actually RetroArch only displays what the core tells it to display.
Crop Overscan is implemented by the core itself. What would need to happen is that the core should set_geometry depending on the required resolution.

**Comment by fr500** [comment]
Closing this for now, I think the crop overscan option will be deprecated anyway.
We should investigate on the cores themselves
